drivers/ethernet/eth_stm32_hal.c

```diff
@@ -932,3 +1127,293 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		(clock_control_subsys_t *)&eth_cfg->pclken,
+#else
+		(clock_control_subsys_t *)&eth_cfg->pclken_ptp,
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+		&ptp_hclk_rate);
+	if (ret) {
+		LOG_ERR("Failed to query ethernet clock");
+		return -EIO;
+	}
+
+	/* Program the subsecond increment register based on the PTP clock freq */
+	if (NSEC_PER_SEC % CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ != 0) {
+		LOG_ERR("PTP clock period must be an integer nanosecond value");
+		return -EINVAL;
+	}
+	ss_incr_ns = NSEC_PER_SEC / CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ;
+	if (ss_incr_ns > UINT8_MAX) {
+		LOG_ERR("PTP clock period is more than %d nanoseconds", UINT8_MAX);
+		return -EINVAL;
+	}
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACSSIR = ss_incr_ns << ETH_MACMACSSIR_SSINC_Pos;
+#else
+	heth->Instance->PTPSSIR = ss_incr_ns;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Program timestamp addend register */
+	eth_dev_data->clk_ratio =
+		((double)CONFIG_ETH_STM32_HAL_PTP_CLOCK_SRC_HZ) / ((double)ptp_hclk_rate);
+	/*
+	 * clk_ratio is a ratio between desired PTP clock frequency and HCLK rate.
+	 * Because HCLK is defined by a physical oscillator, it might drift due
+	 * to manufacturing tolerances and environmental effects (e.g. temperature).
+	 * clk_ratio_adj compensates for such inaccuracies. It starts off as 1.0
+	 * and gets adjusted by calling ptp_clock_stm32_rate_adjust().
+	 */
+	eth_dev_data->clk_ratio_adj = 1.0f;
+	addend_val =
+		UINT32_MAX * eth_dev_data->clk_ratio * eth_dev_data->clk_ratio_adj;
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSAR = addend_val;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSADDREG;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSADDREG_Msk) {
+		k_yield();
+	}
+#else
+	heth->Instance->PTPTSAR = addend_val;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSARU;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSARU_Msk) {
+		k_yield();
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Enable fine timestamp correction method */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSCFUPDT;
+#else
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSFCU;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Enable nanosecond rollover into a new second */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSCTRLSSR;
+#else
+	heth->Instance->PTPTSCR |= ETH_PTPTSSR_TSSSR;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Initialize timestamp */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACSTSUR = 0;
+	heth->Instance->MACSTNUR = 0;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSINIT;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSINIT_Msk) {
+		k_yield();
+	}
+#else
+	heth->Instance->PTPTSHUR = 0;
+	heth->Instance->PTPTSLUR = 0;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSSTI;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSSTI_Msk) {
+		k_yield();
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```

Comment:
Nit , I wonder if one big #if #else wouldn't be easier to read.

Reply:
I'd prefer to keep like this for now because steps across MCUs perfectly align, but registers don't. This makes it easier to make sure the right code is still executed. Can be revised if we find STM family that follows a different procedure that we want to support.